_posts/2019-12-19-aad-pod-binding-sql-server-conn.md

```diff
@@ -0,0 +1,226 @@
+
+After getting better understanding about what managed identities are and how we can create Azure AD based contained database user in our SQL Database, we should somehow assign our identity to our application. But how? 
+
+If you read the [official document](https://docs.microsoft.com/en-us/azure/app-service/app-service-web-tutorial-connect-msi#modify-aspnet-core), you can see Azure App Service is used. And the assignment is performed with this simple az cli command:
+
+```
+az webapp identity assign --resource-group myResourceGroup --name <myAppName>
+```
+
+After running this command, Azure Resource Manager configures the identity on the underlying VM and updates the Azure Instance Metadata Service Identity endpoint with the assigned managed identity service principal client ID and certificate. And the code that is running on the VM can request a token from Azure Instance Metadata Service identity endpoint, accessible only from within the VM: http://169.254.169.254/metadata/identity/oauth2/token.
+
+However for our scenario, we are not using Azure App Service. We just deploy our `.NET Core` application to the `k8s` cluster. So we don't have such an option to directly assign our application to user assigned managed identity. When we deploy our application to the cluster, somehow we should be able to assign these identities in the pod level. Fortunately, [AAD Pod Identity](https://github.com/Azure/aad-pod-identity) is used for this purpose. It enables `k8s` applications to access cloud resources securely with Azure AD. 
+
+[This article](https://medium.com/microsoftazure/pod-identity-5bc0ffb7ebe7) is very nice to get an idea of AAD Pod Identity concept. Basically, when a pod is scheduled to a node, `aad-pod-identity` ensures that a pre-configured user assigned identity is assigned to the underlying VM. If you follow the [Getting Started](https://github.com/Azure/aad-pod-identity#getting-started) steps, you can easily setup your application in cluster with identity binding.
+Of course, before starting you have to have a running `k8s` cluster. 
+
+First we need to enable AAD Pod Identity in our cluster by deploying it. It includes 1 `NMI` (Node Managed Identity) daemon set and 2 `MIC` (Managed ıdentity Controller) pods and several custom resources. In order to get better insights about this plugin, you can read [concept](https://github.com/Azure/aad-pod-identity/blob/master/docs/design/concept.md) and investigate the [concept diagram](https://github.com/Azure/aad-pod-identity/blob/master/docs/design/concept.png).
+
+Since our cluster is non-RBAC, deploy with the following command:
+
+```
+kubectl apply -f https://raw.githubusercontent.com/Azure/aad-pod-identity/master/deploy/infra/deployment.yaml
+```
+
+Create User Assigned Managed Identity and note the `clientId` from output for later use:
+
+```
+az identity create -g myResourceGroup -n myidentity -o json
+```
+
+Install created identity to our cluster by deploying the following:
+
+```
```

Comment:
Add `yaml` for syntax highlighting.

<img width="464" alt="image" src="https://user-images.githubusercontent.com/1406759/72078533-9538e200-32f9-11ea-91f5-c3a7791dceb7.png">

<img width="582" alt="image" src="https://user-images.githubusercontent.com/1406759/72078555-9ec24a00-32f9-11ea-9fe1-943997f568c9.png">